File: problems/0027-remove-element/solution.js
/**
 * @problemNumber: 27
 * @problemName: Remove Element
 * @difficulty: Easy
 * @topics: Array, Two Pointers
 * @companies: Facebook, Amazon, Google
 */

/**
 * Approach: Two Pointers
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 *
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
function removeElement(nums, val) {
    let k = 0;
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] !== val) {
            nums[k] = nums[i];
            k++;
        }
    }
    return k;
}

// Test cases
let nums1 = [3, 2, 2, 3];
let k1 = removeElement(nums1, 3);
console.log(k1, nums1.slice(0, k1)); // 2, [2,2]

let nums2 = [0, 1, 2, 2, 3, 0, 4, 2];
let k2 = removeElement(nums2, 2);
console.log(k2, nums2.slice(0, k2)); // 5, [0,1,3,0,4] (order may vary)

module.exports = { removeElement };

File: problems/0027-remove-element/solution.py
"""
Problem Number: 27
Problem Name: Remove Element
Difficulty: Easy
Topics: Array, Two Pointers
Companies: Facebook, Amazon, Google
"""

from typing import List

class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        """
        Approach: Two Pointers
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        k = 0
        for i in range(len(nums)):
            if nums[i] != val:
                nums[k] = nums[i]
                k += 1
        return k

# Test cases
if __name__ == "__main__":
    solution = Solution()
    nums1 = [3,2,2,3]
    k1 = solution.removeElement(nums1, 3)
    print(k1, nums1[:k1]) # 2, [2,2]

    nums2 = [0,1,2,2,3,0,4,2]
    k2 = solution.removeElement(nums2, 2)
    print(k2, nums2[:k2]) # 5, [0,1,3,0,4]
